RecDP/tests/test_llmutils_pipelines.py

```diff
@@ -184,3 +184,25 @@ def test_pipeline_execute_pandasdf_spark(self):
         ret = pipeline.execute(df)
         display(ret.toPandas())
 
+    def test_llm_rag_pipeline(self):
```

Comment:
proposed pipeline is looking good to me, let's align with the plan firstly. It is quite different to what we discussed in the meeting I think.

Reply:
If we directly call the function of langchain in the pipeline of recdp, and then automatically  perform data conversion for the function of langchain in the pipeline, it would be easier to use. Will this implementation have some significant gap?

```
from langchain.document_loaders import TextLoader 
from pyrecdp.primitives.operations import TextFix 
pipeline = TextPipeline() 
ops = [ 
    TextLoader("./index.html"), 
    TextFix(text_type='html'), 
    RecursiveCharacterTextSplitter(chunk_size=500,  chunk_overlap=0) 
    """ 
    Some other ops. 
    """ 
    ), 
]  
pipeline.add_operations(ops) 
```

Reply:
Previously, we discussed not using LangChain. Did you mean the PDF reader? At the time, we felt it was a bit heavy, and our reader can support one text per document.

The pipeline I am proposing now is to divide it into three operators following the Rag workflow: reader, split, and embedding.

The implementation of each operator is actually a wrapper of third-party readers, splits, and embeddings. Currently, we can use LangChain or LlamaIndex.

Regarding vecstore extensibility, I saw that LangChain does not have a unified implementation for persistence. Therefore, I made an abstraction here. Currently, the LangChain design requires us to handle the persistence of different vecstored separately.

However, if we change to LlamaIndex, we can unify the persistence using the vecstore.persist API of LlamaIndex.